### PR TITLE
ARROW-1618: [JAVA] Reduce Heap Usage (Phase 1)

### PR DESCRIPTION
**Pull request by siddharthteotia**
cc @jacques-n , @icexelloss , @BryanCutler 

This is initial small phase of our attempt to reduce heap usage per vector.

As part of investigation, we realized its better to address few things as part of subtasks for ARROW-1463. Accordingly, I need to update the requirements document w.r.t heap usage for ARROW-1471.

This patch gets rid of Release Listener object in Allocation Manager as all the logic is implemented as part of AllocationManager itself.

https://docs.google.com/document/d/1MU-ah_bBHIxXNrd7SkwewGCOOexkXJ7cgKaCis5f-PI/edit

### CODE REVIEW

File: java/memory/src/main/java/org/apache/arrow/memory/AllocationManager.java [hunk 1]
```diff
@@ -35,6 +35,7 @@
 import io.netty.buffer.ArrowBuf;
 import io.netty.buffer.PooledByteBufAllocatorL;
 import io.netty.buffer.UnsafeDirectLittleEndian;
+import sun.misc.Unsafe;
```
Comment (jacques-n) [inline java/memory/src/main/java/org/apache/arrow/memory/AllocationManager.java]: why added?

File: java/memory/src/main/java/org/apache/arrow/memory/AllocationManager.java [hunk 2]
```diff
@@ -151,54 +152,40 @@ private BufferLedger associate(final BaseAllocator allocator, final boolean reta
...
-    public ReleaseListener(BufferAllocator allocator) {
-      this.allocator = allocator;
-    }
-
-    /**
-     * Can only be called when you already hold the writeLock.
-     */
-    public void release() {
-      allocator.assertOpen();
+  private void release(BufferAllocator allocator) {
```
Comment (jacques-n) [inline java/memory/src/main/java/org/apache/arrow/memory/AllocationManager.java]: Maybe use BufferLedger here so we ensure no misuse of this?

File: java/memory/src/main/java/org/apache/arrow/memory/AllocationManager.java (hunk 1 above)
Comment (siddharthteotia) [reply java/memory/src/main/java/org/apache/arrow/memory/AllocationManager.java]: My bad :( result of auto imports based on some code changes that were later removed.

**Comment by siddharthteotia** [comment]
addressed review comments.

**Comment by jacques-n** [comment]
lgtm +1

**Comment by wesm** [comment]
Cool, can merge after the build runs

**Comment by siddharthteotia** [comment]
This can be merged

**Comment by wesm** [comment]
+1

**Pull request closed without merging**